File: 2trabajoConstructores/go/interfaz/main.go
package main

import "fmt"

func main() {

	fmt.Println("=== INTERFAZ EN GO ===")

	// Declaraciones usando la interfaz
	var animal IAnimal

	fmt.Println("\n=== SIN PARÁMETROS ===")
	animal = NuevoAnimalSinParametros()
	animal.HacerSonido()

	fmt.Println("\n=== CON PARÁMETROS ===")
	animal = NuevoAnimalConParametros("Perro", "Guau")
	animal.HacerSonido()

	fmt.Println("\n=== SOBRECARGADO ===")
	animal = NuevoAnimalVacio()
	animal.HacerSonido()

	animal = NuevoAnimalSoloNombre("Gato")
	animal.HacerSonido()

	animal = NuevoAnimalCompleto("Vaca", "Muuu")
	animal.HacerSonido()

	fmt.Println("\n=== COPIA ===")
	original := NuevoAnimalConParametros("Loro", "Hola")
	copia := NuevoAnimalCopia(original)
	copia.HacerSonido()
}

File: 2trabajoConstructores/js/interfaz/main.js
const Perro = require("./Perro");
const Gato = require("./Gato");
const Loro = require("./Loro");
const AnimalCopia = require("./AnimalCopia");

console.log("=== PERRO ===");
let p = new Perro("Firulais");
p.hacerSonido();

console.log("\n=== GATO ===");
let g = new Gato("Misifú");
g.hacerSonido();

console.log("\n=== LORO ===");
let l = new Loro("Paco", "¡Hola humano!");
l.hacerSonido();

console.log("\n=== COPIA ===");
let copia = new AnimalCopia(p);
copia.hacerSonido();
